apps/dashboard/app/models/posix_file.rb

```diff
@@ -79,17 +81,25 @@ def files_to_zip
   end
 
   def directory?
-    path.stat.directory?
+    stat.directory?
   end
 
   def ls
     path.each_child.map do |child_path|
-      PosixFile.stat(child_path)
-    end.select do |stats|
-      valid_encoding = stats[:name].to_s.valid_encoding?
-      Rails.logger.warn("Not showing file '#{stats[:name]}' because it is not a UTF-8 filename.") unless valid_encoding
-      valid_encoding
-    end.sort_by { |p| p[:directory] ? 0 : 1 }
+      PosixFile.new(child_path)
+    end.select(&:valid?)
```

Comment:
Really most of these changes come from my wanting to do this: `select(&:valid?)`. I figured a `self#valid?` API is better than a `valid?(stat)` (where `stat` is just a hash from `PosixFile.stat`)